File: frontend/src/lib/tokenize.ts
export interface Token {
  type: "word" | "other";
  text: string;
}

const WORD_REGEX = /[\p{L}]+(?:[''\-][\p{L}]+)*/gu;

export function tokenize(text: string): Token[] {
  const tokens: Token[] = [];
  let lastIndex = 0;

  for (const match of text.matchAll(WORD_REGEX)) {
    const start = match.index!;
    const end = start + match[0].length;

    if (start > lastIndex) {
      tokens.push({ type: "other", text: text.slice(lastIndex, start) });
    }

    tokens.push({ type: "word", text: match[0] });
    lastIndex = end;
  }

  if (lastIndex < text.length) {
    tokens.push({ type: "other", text: text.slice(lastIndex) });
  }

  return tokens;
}

File: app/lib/tokenize.py
import regex

_WORD_REGEX = regex.compile(r"[\p{L}]+(?:[''\-][\p{L}]+)*")


def word_tokenize(text: str) -> list[str]:
    """
    Split `text` into word tokens.

    A token is a run of Unicode letters, optionally containing internal
    apostrophes (straight or curly) or hyphens between letter runs — e.g.
    "don't", "well-known", "Müller". Digits, punctuation, and whitespace
    are skipped, and leading/trailing apostrophes/hyphens are not included.

    Mirrors the frontend tokenizer in frontend/src/lib/tokenize.ts so the
    backend and UI agree on what counts as a word.
    """
    return [m.group(0) for m in _WORD_REGEX.finditer(text)]
